packages/node/src/core/ethereum/authorization/authorization-application.ts

```diff
@@ -0,0 +1,101 @@
+import get from 'lodash/get';
+import flatMap from 'lodash/flatMap';
+import isNil from 'lodash/isNil';
+import * as logger from '../../utils/logger';
+import {
+  ApiCall,
+  AuthorizationByEndpointId,
+  ClientRequest,
+  LogsErrorData,
+  RequestErrorCode,
+  RequestStatus,
+  WalletDataByIndex,
+} from '../../../types';
+
+function mapApiCalls(
+  apiCalls: ClientRequest<ApiCall>[],
+  authorizationsByEndpointId: AuthorizationByEndpointId
+): LogsErrorData<ClientRequest<ApiCall>[]> {
+  const logsWithApiCalls: LogsErrorData<ClientRequest<ApiCall>>[] = apiCalls.map((apiCall) => {
+    // Don't overwrite any existing error codes or statuses
+    if (apiCall.errorCode || apiCall.status !== RequestStatus.Pending) {
+      return [[], null, apiCall];
+    }
+
+    // There should always be an endpointId at this point, but just in case, check again
+    // and drop the request if it is missing. If endpointId is missing, it means that the
+    // template was not loaded
+    if (!apiCall.endpointId) {
+      const log = logger.pend('ERROR', `No endpoint ID found for Request ID:${apiCall.id}`);
+      const updatedApiCall = {
+        ...apiCall,
+        status: RequestStatus.Blocked,
+        errorCode: RequestErrorCode.TemplateNotFound,
+      };
+      return [[log], null, updatedApiCall];
+    }
+
+    const isRequestedAuthorized = get(authorizationsByEndpointId, [apiCall.endpointId, apiCall.requesterAddress]);
+
+    // If we couldn't fetch the authorization status, block the request until the next run
+    if (isNil(isRequestedAuthorized)) {
+      const log = logger.pend('WARN', `Authorization not found for Request ID:${apiCall.id}`);
+      const updatedApiCall = {
+        ...apiCall,
+        status: RequestStatus.Blocked,
+        errorCode: RequestErrorCode.AuthorizationNotFound,
+      };
+      return [[log], null, updatedApiCall];
+    }
+
+    if (isRequestedAuthorized) {
+      return [[], null, apiCall];
+    }
+
+    const log = logger.pend(
+      'WARN',
+      `Client:${apiCall.requesterAddress} is not authorized to access Endpoint ID:${apiCall.endpointId} for Request ID:${apiCall.id}`
+    );
+    const updatedApiCall = {
+      ...apiCall,
+      status: RequestStatus.Errored,
```

Comment:
Is it correct that a request is set to "Errored" if the authorization returns false? This will result in an error transaction at the end. Can a malicious user drain someone else's wallet by spamming unauthorised requests?

Reply:
Yeah but does that stop a malicious client from spamming unauthorized requests? Since they all get set to `Errored` they still result in `error()` transactions using the requester's wallet, right? Maybe I'm misunderstanding that part

Reply:
I think you're imagining a use case where the dapp looks like this
https://clc-group.github.io/workshop-honeycomb-react/
where the dapp owner pays for the gas but the user freely triggers however many requests they want. In reality, the client (i.e. the dapp) would be implemented in a way that either makes a set request that cannot be parametrized by the user or the user puts down an ETH deposit that covers the gas fee of the request + possibly more etc.